// dmmd.js/dmath/exports.js
"use strict";

// Imports
const functions = {
    absolute: require("./functions/absolute.js"),
    get abs() { return functions.absolute },

	acos: require("./functions/acos.js"),

	acosh: require("./functions/acosh.js"),

	asin: require("./functions/asin.js"),

	asinh: require("./functions/asinh.js"),

	atan: require("./functions/atan.js"),

	atan2: require("./functions/atan2.js"),

	atanh: require("./functions/atanh.js"),

	cbrt: require("./functions/cbrt.js"),

    ceil: require("./functions/ceil.js"),

	clz32: require("./functions/clz32.js"),

	cos: require("./functions/cos.js"),

	cosh: require("./functions/cosh.js"),

	e: require("./functions/e.js"),

	exp: require("./functions/exp.js"),

	expm1: require("./functions/expm1.js"),

    floor: require("./functions/floor.js"),

	fround: require("./functions/fround.js"),

	hypot: require("./functions/hypot.js"),

	imul: require("./functions/imul.js"),

	ln2: require("./functions/ln2.js"),

	ln2e: require("./functions/ln2e.js"),

	ln10: require("./functions/ln10.js"),

	ln10e: require("./functions/ln10e.js"),

	log: require("./functions/log.js"),

	log1p: require("./functions/log1p.js"),

	log2: require("./functions/log2.js"),

	log10: require("./functions/log10.js"),

	max: require("./functions/max.js"),

	min: require("./functions/min.js"),
    
    negative: require("./functions/negative.js"),
    get neg() { return functions.negative },

	pi: require("./functions/pi.js"),

	pow: require("./functions/pow.js"),
    
	random: require("./functions/random.js"),
    
	round: require("./functions/round.js"),

	sign: require("./functions/sign.js"),

	sin: require("./functions/sin.js"),

	sinh: require("./functions/sinh.js"),

	sqrt: require("./functions/sqrt.js"),

	sqrt1_2: require("./functions/sqrt1_2.js"),

	sqrt2: require("./functions/sqrt2.js"),

	tan: require("./functions/tan.js"),

	tanh: require("./functions/tanh.js"),
    
	trunc: require("./functions/trunc.js"),
};
const operators = {
    divide: require("./operators/divide.js"),

    modulo: require("./operators/modulo.js"),
    get remainder() { return operators.modulo },

    multiply: require("./operators/multiply.js"),
    get times() { return operators.multiply },

    plus: require("./operators/plus.js"),
    get add() { return operators.plus },

    subtract: require("./operators/subtract.js"),
    get minus() { return operators.subtract }
};
const utils = {
    dgroup: require("./utils/dgroup.js"),

    format: require("./utils/format.js"),

    parse: require("./utils/parse.js"),

    shift: require("./utils/shift.js")
};

// Exports
module.exports = {
    functions, ...functions,
    operators, ...operators,
    utils, ...utils
};
