// dmmd.js/dpaint/exports.js
"use strict";

// Imports
const colors = require("./basics/colors.js");
const customs = require("./basics/customs.js");
const styles = require("./basics/styles.js");
const advanced = {
    cmyk: require("./advanced/cmyk.js"),

    decimal: require("./advanced/decimal.js"),
    get dec() { return advanced.decimal },

    eightBit: require("./advanced/eightBit.js"),

    hex: require("./advanced/hex.js"),

    hex3: require("./advanced/hex3.js"),

    hsl: require("./advanced/hsl.js"),

    hsv: require("./advanced/hsv.js"),

    preset: require("./advanced/preset.js"),

    rgb: require("./advanced/rgb.js")
};
const basics = {
    colors, ...colors,
    customs, ...customs,
    styles, ...styles
};
const utils = {
    bColor: require("./utils/bColor.js"),

    bStyle: require("./utils/bStyle.js"),

    convert: require("./utils/convert.js"),

    parse: require("./utils/parse.js"),

    tag: require("./utils/tag.js")
};

// Exports
module.exports = {
    advanced, ...advanced,
    basics, ...basics,
    utils, ...utils
};
