// dmmd.js/dpaint/basics/colors.js
"use strict";

// Imports
const bColor = require("../utils/bColor.js");

// Variables
let colors = {
    black: (s, bg) => bColor("30", "40", s, bg),

    red: (s, bg) => bColor("31", "41", s, bg),

    green: (s, bg) => bColor("32", "42", s, bg),

    yellow: (s, bg) => bColor("33", "43", s, bg),

    blue: (s, bg) => bColor("34", "44", s, bg),

    magenta: (s, bg) => bColor("35", "45", s, bg),

    cyan: (s, bg) => bColor("36", "46", s, bg),

    white: (s, bg) => bColor("37", "47", s, bg),

    brightBlack: (s, bg) => bColor("90", "100", s, bg),
    get gray() { return colors.brightBlack },
    get grey() { return colors.brightBlack },

    brightRed: (s, bg) => bColor("91", "101", s, bg),

    brightGreen: (s, bg) => bColor("92", "102", s, bg),

    brightYellow: (s, bg) => bColor("93", "103", s, bg),

    brightBlue: (s, bg) => bColor("94", "104", s, bg),

    brightMagenta: (s, bg) => bColor("95", "105", s, bg),

    brightCyan: (s, bg) => bColor("96", "106", s, bg),

    brightWhite: (s, bg) => bColor("97", "107", s, bg)
};

// Exports
module.exports = colors;
