// dmmd.js/dpaint/basics/styles.js
"use strict";

// Imports
const bStyle = require("../utils/bStyle.js");

// Variables
let styles = {
    reset: () => bStyle(null, "0", ""),

    bold: s => bStyle("1", "22", s),
    unbold: () => bStyle(null, "22", ""),

    faint: s => bStyle("2", "22", s),
    get dim() { return styles.faint },
    unfaint: () => bStyle(null, "22", ""),
    get undim() { return styles.unfaint },

    italic: s => bStyle("3", "23", s),
    unitalic: () => bStyle(null, "23", ""),

    underline: s => bStyle("4", "24", s),
    ununderline: () => bStyle(null, "24", ""),

    blink: s => bStyle("5", "25", s),
    get flash() { return styles.blink },
    unblink: () => bStyle(null, "25", ""),
    get unflash() { return styles.unblink },

    rapidBlink: s => bStyle("6", "25", s),
    get rapidFlash() { return styles.rapidBlink },
    unrapidBlink: () => bStyle(null, "25", ""),
    get unrapidFlash() { return styles.unrapidBlink },

    revert: s => bStyle("7", "27", s),
    get invert() { return styles.revert },
    unrevert: () => bStyle(null, "27", ""),
    get uninvert() { return styles.unrevert },

    conceal: s => bStyle("8", "28", s),
    get hide() { return styles.conceal },
    unconceal: () => bStyle(null, "28", ""),
    get unhide() { return styles.unconceal },

    crossOut: s => bStyle("9", "29", s),
    get strike() { return styles.crossOut },
    get strikeThrough() { return styles.crossOut },
    uncrossOut: () => bStyle(null, "29", ""),
    get unstrike() { return styles.uncrossOut },
    get unstrikeThrough() { return styles.uncrossOut },

    doubleUnderline: s => bStyle("21", "24", s),
    undoubleUnderline: () => bStyle(null, "24", ""),

    overline: s => bStyle("53", "55", s),
    unoverline: () => bStyle(null, "55", ""),
};

// Exports
module.exports = styles;
